cabal-validate/src/Main.hs

```diff
@@ -0,0 +1,430 @@
+  withTiming opts title action
+  T.putStrLn ""
+
+-- | Compiler with version number like @ghc-9.6.6@.
+baseHc :: Opts -> FilePath
+baseHc opts = "ghc-" <> showVersion (compilerVersion $ compiler opts)
+
+-- | Base build directory for @cabal-validate@.
+baseBuildDir :: Opts -> FilePath
+baseBuildDir opts = "dist-newstyle-validate-" <> baseHc opts
+
+-- | Absolute path to the build directory for this architecture.
+--
+-- This is a path nested fairly deeply under `baseBuildDir`.
+buildDir :: Opts -> FilePath
+buildDir opts =
+  cwd opts
+    </> baseBuildDir opts
+    </> "build"
+    </> archPath opts
+    </> baseHc opts
+
+-- | @--num-threads@ argument for test suites.
+--
+-- This isn't always used because some test suites are finicky and only accept
+-- @-j@.
+jobsArgs :: Opts -> [String]
+jobsArgs opts = ["--num-threads", show $ jobs opts]
+
+-- | Default arguments for invoking @cabal@.
+cabalArgs :: Opts -> [String]
+cabalArgs opts =
+  [ "--jobs=" <> show (jobs opts)
+  , "--with-compiler=" <> compilerExecutable (compiler opts)
+  , "--builddir=" <> baseBuildDir opts
+  , "--project-file=" <> projectFile opts
+  ]
+
+-- | The `buildDir` for @cabal-testsuite-3@.
+cabalTestsuiteBuildDir :: Opts -> FilePath
+cabalTestsuiteBuildDir opts =
+  buildDir opts
+    </> "cabal-testsuite-3"
+
+-- | Arguments for @cabal build@.
+cabalNewBuildArgs :: Opts -> [String]
+cabalNewBuildArgs opts = "build" : cabalArgs opts
+
+-- | Arguments for @cabal list-bin@.
+--
+-- This is used to find the binaries for various test suites.
+cabalListBinArgs :: Opts -> [String]
+cabalListBinArgs opts = "list-bin" : cabalArgs opts
+
+-- | Get the binary for a given @cabal@ target by running @cabal list-bin@.
+cabalListBin :: Opts -> String -> IO FilePath
+cabalListBin opts target = do
+  let args = cabalListBinArgs opts ++ [target]
+  stdout <-
+    readProcessStdout_ $
+      proc (cabal opts) args
+
+  pure (T.unpack $ T.strip $ T.toStrict $ T.decodeUtf8 stdout)
+
+-- | Get the RTS arguments for invoking test suites.
+--
+-- These seem to only be used for some of the test suites, I'm not sure why.
+rtsArgs :: Opts -> [String]
+rtsArgs opts =
+  case archPath opts of
+    "x86_64-windows" ->
+      -- See: https://github.com/haskell/cabal/issues/9571
+      if compilerVersion (compiler opts) > makeVersion [9, 0, 2]
+        then ["+RTS", "--io-manager=native", "-RTS"]
+        else []
+    _ -> []
+
+-- | Run a binary built by @cabal@ and output timing information.
+--
+-- This is used to run many of the test suites.
+timedCabalBin :: Opts -> String -> String -> [String] -> IO ()
+timedCabalBin opts package component args = do
+  command <- cabalListBin opts (package <> ":" <> component)
+  timedWithCwd
+    opts
+    package
+    command
+    args
+
+-- | Print the configuration for CI logs.
+printConfig :: Opts -> IO ()
+printConfig opts = do
+  putStrLn $
+    "compiler:          "
+      <> compilerExecutable (compiler opts)
+      <> "\ncabal-install:     "
```

Comment:
This is purely stylistic and subjective, but I’d much rather not read `\n`. Luckily, `unlines` does the job. (As a bonus, half of `<>` will turn into `,` — what a profit we get if we were paid by how much we’re saving on characters!..)